core/thread/inc/TPool.h

```diff
@@ -0,0 +1,204 @@
+   // //and a compile-time check that the argument list implements a front() method (all STL sequence containers have it)
+   template<class F> auto Map(F func, unsigned nTimes) -> std::vector<decltype(func())>;
+   template<class F, class T> auto Map(F func, T &args) -> std::vector < decltype(++(args.begin()), args.end(), func(args.front())) >;
+   // /// \cond doxygen should ignore these methods
+   template<class F, class INTEGER> auto Map(F func, ROOT::TSeq<INTEGER> args) -> std::vector<decltype(func(args.front()))>;
+   template<class F> TObjArray Map(F func, TCollection &args);
+   template<class F, class T> auto Map(F func, std::initializer_list<T> args) -> std::vector<decltype(func(*args.begin()))>;
+   template<class F, class T> auto Map(F func, std::vector<T> &args) -> std::vector<decltype(func(args.front()))>;
+   // // // / \endcond
+
+   // // MapReduce
+   // // the late return types also check at compile-time whether redfunc is compatible with func,
+   // // other than checking that func is compatible with the type of arguments.
+   // // a static_assert check in TPool<subc>::Reduce is used to check that redfunc is compatible with the type returned by func
+   template<class F, class T, class R> auto MapReduce(F func, T*args, int nEvents, R redfunc) -> decltype(func(args, 0));
+   template<class F, class R> auto MapReduce(F func, unsigned nTimes, R redfunc) -> decltype(func());
+   template<class F, class INTEGER, class R> auto MapReduce(F func, ROOT::TSeq<INTEGER> args, R redfunc) -> decltype(func(args.front()));
+   template<class F, class T, class R> auto MapReduce(F func, T &args, R redfunc) -> decltype(++(args.begin()), args.end(), func(args.front()));
+   // /// \cond doxygen should ignore these methods
+   template<class F, class R> auto MapReduce(F func, TCollection &args, R redfunc) -> decltype(func(nullptr));
+   template<class F, class T, class R> auto MapReduce(F func, std::initializer_list<T> args, R redfunc) -> decltype(func(*args.begin()));
+   template<class F, class T, class R> auto MapReduce(F func, std::vector<T> &args, R redfunc) -> decltype(func(args.front()));
+   // /// \endcond
+
+protected:
+   template<class T, class R> auto Reduce(const std::vector<T> &objs, R redfunc) -> decltype(redfunc(objs));
+   template<class T, class BINARYOP> auto Reduce(const std::vector<T> &objs, BINARYOP redfunc) -> decltype(redfunc(objs.front(), objs.front()));
+};
+
+//////////////////////////////////////////////////////////////////////////
+/// Execute func (with no arguments) nTimes in parallel.
+/// A vector containg executions' results is returned.
+/// Functions that take more than zero arguments can be executed (with
+/// fixed arguments) by wrapping them in a lambda or with std::bind.
+template<class subc> template<class F>
+auto TPool<subc>::Map(F func, unsigned nTimes) -> std::vector<decltype(func())>
+{
+   return static_cast<subc*>(this)->Map(func, nTimes);
```

Comment:
It's fairly common to factor the `static_cast<subc*>(this)` out into an inline function called e.g. `Derived()`, returning `*static_cast<subc*>(this)` (note the deref).